Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp

```diff
@@ -0,0 +1,822 @@
+        DFG_ASSERT(m_graph, header->at(0), header->predecessors.size() > 1, header->predecessors.size());
+
+        uint32_t preHeaderCount = 0;
+        for (uint32_t i = header->predecessors.size(); i--;) {
+            BasicBlock* predecessor = header->predecessors[i];
+            if (m_graph.m_cpsDominators->dominates(header, predecessor))
+                continue;
+
+            preHeader = predecessor;
+            ++preHeaderCount;
+        }
+
+        if (preHeaderCount != 1)
+            return false;
+
+        data.preHeader = preHeader;
+        return true;
+    }
+
+    bool locateTail(LoopData& data)
+    {
+        BasicBlock* header = data.header();
+        BasicBlock* tail = nullptr;
+
+        for (BasicBlock* predecessor : header->predecessors) {
+            if (!m_graph.m_cpsDominators->dominates(header, predecessor))
+                continue;
+
+            if (tail) {
+                dataLogLnIf(verbose, "Skipping loop with header ", *header, " since it contains two tails: ", *predecessor, " and ", *tail);
+                return false;
+            }
+
+            tail = predecessor;
+        }
+
+        if (!tail) {
+            dataLogLnIf(verbose, "Skipping loop with header ", *header, " since it has no tail");
+            return false;
+        }
+
+        // PreHeader                          PreHeader
+        //  |                                  |
+        // Header <---                        Header_0
+        //  |        |       unrolled to       |
+        //  |       Tail  =================>  Branch_0
+        //  |        |                         |
+        // Branch ----                        Tail_0
+        //  |                                  |
+        // Next                               ...
+        //                                     |
+        //                                    Header_n
+        //                                     |
+        //                                    Branch_n
+        //                                     |
+        //                                    Next
+        //
+        // FIXME: This is not supported yet. We should do it only if it's profitable.
+        if (!tail->terminal()->isBranch()) {
+            dataLogLnIf(verbose, "Skipping loop with header ", *header, " since it has a non-branch tail");
+            return false;
+        }
+
+        ASSERT(tail->terminal()->isBranch() && tail->successors().size() == 2);
+        for (BasicBlock* successor : tail->successors()) {
+            if (data.loop->contains(successor))
+                continue;
+            data.next = successor;
+        }
+        data.tail = tail;
+        return true;
+    }
+
+    bool isSupportedConditionOp(NodeType op);
+    bool isSupportedUpdateOp(NodeType op);
+    std::function<bool(CheckedInt32, CheckedInt32)> comparisonFunction(Node* condition);
+    std::function<CheckedInt32(CheckedInt32, CheckedInt32)> updateFunction(Node* update);
```

Comment:
Let's avoid using `std::function`. This allocates object in system malloc and not efficient. In this case, we should just use function pointer.